gateway_to_backend/README.md

```diff
@@ -0,0 +1,579 @@
+
+## API description
+
+This section introduces the API's MQTT topics and its contents and is 
+split in several subsections, one for each module:
+
+    1. [Gateway status module](#gateway-status-module)
+    1. [Config module](#config-module)
+    1. [Data module](#data-module)
+    1. [OTAP module](#otap-module)
+
+Each subsection contains the relevant topic and the payload carried
+by it. This sections point out the specific protocol buffer message
+that must be implemented within the common 
+[Generic Message][message_GenericMessage]. Please refer to 
+[API extension](#api-extension) for further details.
+
+In MQTT, the QoS is an important concept. Most of the gateway to broker
+communication should happen with QoS2 to be sure the messages are
+received or published one and only one time. However, as many packets
+can be received from a network and multiple gateways can publish on the
+same broker, message received event should be published with QoS1.
+Depending on the quality of the connection between gateway and Backends,
+it can create duplicates that will have to be filtered by the backends.
+To ease this filtering, each event contains a unique identifier (random
+64 bits value) in its header.
+
+### Gateway status module
+
+This module oversees the discovery of gateways for the different
+backends. It also allows the backends to monitor the gateways status.
+
+#### Status message
+
+> **topic:** gw-event/status/*\<gw_id\>*
+>
+> **content:** [StatusEvent message format][message_StatusEvent]
+
+The message must have the retained flag set to true to allow a new
+backend to immediately be noticed of all the available gateways when
+subscribing to this topic.
+
+Once connected to the MQTT broker, status must be set to *ONLINE*.
+
+#### Last will message
+
+Same message with a status set to *OFFLINE* must be set by the gateway as
+a **last will message**.
+
+It will allow backends to easily see which gateway is offline.
+
+#### Keep alive message
+
+This API doesn't enforce the usage of a keep alive message by a gateway
+to avoid additional constraints on the gateway implementation. In fact,
+it implies the generation of a PING request to the broker if there is to
+message generated within a keep alive interval.
+
+Nevertheless, it is recommended to use the keep alive mechanism of MQTT
+to overcome the half-open connection issue of a TCP connection.
+
+The broker will disconnect a client if no message is received within one
+and a half times of the keep alive interval and the last will message
+will be sent. It allows the backends to be notified in a really short
+period of a problem with a gateway connection.
+
+60s is a good value for the keep alive interval.
+
+### Config module
+
+This module oversees configuration of the different sinks attached to
+the gateway.
+
+#### Get configs message
+
+**Request:**
+
+> **topic:** gw-request/get_configs/*\<gw-id\>*
+>
+> **content:** [GetConfigsReq message format][message_GetConfigsReq]
```

Comment:
Link to GenericMessage should be added everywhere. Otherwise it will raise too much question.
Something like:
[Generic message] -> [Wirepas message] -> [GetConfigsReq]

Reply:
Would dots be okay?

`[GenericeMessage][link].[WirepasMessage][link].[GetConfigsReq][link]`

Reply:
Yes sure, it was just to illustrate the idea,. I don't really care about the formatting as long as information is here ;-)

Reply:
Ok! Lets use the dots then